### DISCUSSION TITLE
Benefits of changing StateNotifier<AsyncValue<T>> to AsyncNotifier<T>

### DISCUSSION THREAD

**Discussion by lockieRichter** [rrousselGit/riverpod#2831] [Q&A]
Hi clever folks,

We recently upgraded to the latest Riverpod and Flutter so can now use some of the good new stuff, but I am wondering at the benefit of updating some existing code we have.

Specifically I am wondering what the benefit would be of updating one of our existing `StateNotifier` that uses an `AsyncValue` wrapped around the base object to be an `AsyncNotifier`. 

`StateNotifier<AsyncValue<SomeClass>>` -> `AsyncNotifier<SomeClass>`.

Thanks,
Lockie

**Comment by AhmedLSayed9** [comment]
The benefits I can remember is:
1. Auto handling loading/errors for the build method.
2. `Ref` is a property in the notifier class so you don't have to pass it through the constructor.
3. Only `build` method will re-execute when the provider rebuilds while the notifier class is preserved. This give us the ability to preserve some state across rebuilds.
4. Support the new tags of `.when` such as skipLoading & skipErrors and will support the upcoming features.
5. Support code-generation which will be easy to migrate to meta programming in the future.
6. Easier to get support and help from the community in Discord or Github. If you're asking about `StateNotifier` the first thing you going to hear is "Hey, StateNotifier is legacy. Migrate to Notifier/AsyncNotifier" :)

**Comment by rrousselGit** [comment]
Think of AsyncNotifier as a FutureProvider with ways to modify the state.

It's much simpler to deal with a CRUD, where the `build` method of your AsyncNotifier does a GET, and methods on the notifier do the updates
